### PR TITLE
UT's for Signaling

### PR DESCRIPTION
**Pull request by Varunaditya1**
### Description of changes:

This PR is for improving bugs in the source code and adding additional unit tests to improve coverage, by thoroughly testing various Signaling API's .

The following test cases have been added:
1. `test_signaling_ConstructDescribeSignalingChannelRequest`
2. `test_signaling_ConstructDescribeSignalingChannelRequest_SmallLengthRegion`
3.  `test_signaling_ConstructDescribeSignalingChannelRequest_ChinaRegion`
4.  `test_signaling_ConstructDescribeSignalingChannelRequest_URLOutofMemory`
5.  `test_signaling_ConstructDescribeSignalingChannelRequest_BodyOutofMemory`
6.  `test_signaling_ParseDescribeSignalingChannelResponse_BadParams`
7.  `test_signaling_ParseDescribeSignalingChannelResponse_InvalidJson`
8. `test_signaling_ParseDescribeSignalingChannelResponse_UnexpectedResponse`
9. `test_signaling_ParseDescribeSignalingChannelResponse_InvalidTTL_Low`
10. `test_signaling_ParseDescribeSignalingChannelResponse_InvalidTTL_High`
11. `test_signaling_ParseDescribeSignalingChannelResponse_InvalidChannel_Type`
12. `test_signaling_ParseDescribeSignalingChannelResponse_InvalidChannel_Name`
13. `test_signaling_ParseDescribeSignalingChannelResponse`
14. `test_signaling_ConstructDescribeMediaStorageConfigRequest_BadParams`
15. `test_signaling_ConstructDescribeMediaStorageConfigRequest`
16. `test_signaling_ConstructDescribeMediaStorageConfigRequest_SmallLengthRegion`
17. `test_signaling_ConstructDescribeMediaStorageConfigRequest_ChinaRegion`
18. `test_signaling_ConstructDescribeMediaStorageConfigRequest_URLOutOfMemory`
19. `test_signaling_ConstructDescribeMediaStorageConfigRequest_BodyOutOfMemory`
20. `test_signaling_ParseDescribeMediaStorageConfigResponse_BadParams`
21. `test_signaling_ParseDescribeMediaStorageConfigResponse_InvalidJson`
22. `test_signaling_ParseDescribeMediaStorageConfigResponse_UnexpectedResponse`
23. `test_signaling_ParseDescribeMediaStorageConfigResponse_MissingFields`
24. `test_signaling_ParseDescribeMediaStorageConfigResponse`
25. `test_signaling_ConstructCreateSignalingChannelRequest_BadParams`
26. `test_signaling_ConstructCreateSignalingChannelRequest`
27. `test_signaling_ConstructCreateSignalingChannelRequest_SmallLengthRegion`
28. `test_signaling_ConstructCreateSignalingChannelRequest_ChinaRegion`
29. `test_signaling_ConstructCreateSignalingChannelRequest_WithTags`
30. `test_signaling_ParseCreateSignalingChannelResponse_BadParams`
31. `test_signaling_ParseCreateSignalingChannelResponse_InvalidJson`
32. `test_signaling_ParseCreateSignalingChannelResponse_UnexpectedResponse`
33. `test_signaling_ParseCreateSignalingChannelResponse`
34. `test_signaling_ConstructGetSignalingChannelEndpointRequest_BadParams`
35. `test_signaling_ConstructGetSignalingChannelEndpointRequest`

### Test Steps

```
cmake -S test/unit-test -B build/ -G "Unix Makefiles"  -DBUILD_CLONE_SUBMODULES=ON -DCMAKE_C_FLAGS='--coverage -Wall -Wextra -Werror -DNDEBUG -DLIBRARY_LOG_LEVEL=LOG_DEBUG'
make -C build all
cd build
ctest -E system --output-on-failure
make coverage
```

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: test/unit-test/signaling_api/signaling_api_utest.c [hunk 1]
```diff
@@ -75,3 +75,1168 @@ void test_signaling_ConstructDescribeSignalingChannelRequest_BadParams( void )
...
+
+    result = Signaling_ParseDescribeSignalingChannelResponse( message,
+                                                              messageLength, &( channelInfo ) );
+
+    TEST_ASSERT_EQUAL( SIGNALING_RESULT_INVALID_JSON,
+                       result );
+
+    /* <--------------------------------------------------------------------> */
+
+    const char *message2 = "{"
```
Comment (ActoryOu) [inline test/unit-test/signaling_api/signaling_api_utest.c]: Could you leave a comment here to let us know why message2 is invalid JSON message?

File: test/unit-test/signaling_api/signaling_api_utest.c [hunk 2]
```diff
@@ -75,3 +75,1168 @@ void test_signaling_ConstructDescribeSignalingChannelRequest_BadParams( void )
...
+                                                              messageLength, &( channelInfo ) );
+
+    TEST_ASSERT_EQUAL( SIGNALING_RESULT_UNEXPECTED_RESPONSE,
+                       result );
+}
+
+/*-----------------------------------------------------------*/
+
+/**
+ * @brief Validate Signaling Parse Describe Response functionality for Invalid TTL.
```
Comment (ActoryOu) [inline test/unit-test/signaling_api/signaling_api_utest.c:460]: Add a note here to describe valid TTL range.

File: test/unit-test/signaling_api/signaling_api_utest.c [hunk 3]
```diff
@@ -75,3 +75,1168 @@ void test_signaling_ConstructDescribeSignalingChannelRequest_BadParams( void )
...
+void test_signaling_ParseDescribeMediaStorageConfigResponse( void )
+{
+    SignalingMediaStorageConfig_t mediaStorageConfig;
+    SignalingResult_t result;
+    const char *message  = "{"
+            "\"MediaStorageConfiguration\": {"
+            "\"Status\": \"ENABLED\","
+            "\"StreamARN\": \"arn:aws:kinesisvideo:us-west-2:123456789012:stream/test-stream/1234567890123\""
+            "}"
+            "}";
```
Comment (ActoryOu) [inline test/unit-test/signaling_api/signaling_api_utest.c]:
```suggestion
    const char *message  = \
        "{"
            "\"MediaStorageConfiguration\":
            {"
                "\"Status\": \"ENABLED\","
                "\"StreamARN\": \"arn:aws:kinesisvideo:us-west-2:123456789012:stream/test-stream/1234567890123\""
            "}"
        "}";
```

Is it possible to format this like this?

Comment (Varunaditya1) [reply test/unit-test/signaling_api/signaling_api_utest.c]: Its the uncrustify , i tried to remove it from many places on the JSON format, I think I missed it. Thanks for pointing it out, will do in the upcoming PR.

File: test/unit-test/signaling_api/signaling_api_utest.c (hunk 1 above)
Comment (Varunaditya1) [reply test/unit-test/signaling_api/signaling_api_utest.c]: Done , will be reflected int the next PR. Thanks

File: test/unit-test/signaling_api/signaling_api_utest.c (hunk 2 above)
Comment (Varunaditya1) [reply test/unit-test/signaling_api/signaling_api_utest.c:460]: Done, will be reflected in the next PR. Thanks

**Review by ActoryOu — DISMISSED** [review]

**Review by ActoryOu — DISMISSED** [review]

File: source/signaling_api.c [hunk 4]
```diff
@@ -1237,7 +1233,7 @@ SignalingResult_t Signaling_ConstructJoinStorageSessionRequest( SignalingChannel
             snprintfRetVal = snprintf( pRequestBuffer->pBody,
                                        pRequestBuffer->bodyLength,
                                        "{"
-                                            "\"channelArn\":\"%.*s\""
+                                            "\"ChannelARN\":\"%.*s\""
```
Comment (ActoryOu) [inline source/signaling_api.c]: This must be "[channelArn](https://docs.aws.amazon.com/kinesisvideostreams/latest/dg/API_webrtc_JoinStorageSession.html#KinesisVideo-webrtc_JoinStorageSession-request-channelArn)". Sorry that I misunderstood your question. Please refer to the user manual link.

Comment (ActoryOu) [reply source/signaling_api.c]: Please also update other part unless it's just a variable naming.

File: source/signaling_api.c [hunk 5]
```diff
@@ -1503,14 +1499,15 @@ SignalingResult_t Signaling_ParseWssRecvMessage( const char * pMessage,
...
-    if( pMessage[ messageLength - 1 ] == '\0' )
-    {
-        messageLength--;
-    }
-
     if( result == SIGNALING_RESULT_OK )
     {
+        /* Exclude null terminator in messageLength. */
+        
+        if( pMessage[ messageLength - 1 ] == '\0' )
```
Comment (ActoryOu) [inline source/signaling_api.c]: We should check `messageLength` > 0 before accessing that region of memory.

File: test/unit-test/signaling_api/signaling_api_utest.c [hunk 6]
```diff
@@ -1242,6 +1683,29 @@ void test_signaling_ParseCreateSignalingChannelResponse_UnexpectedResponse( void
...
+        "\"Unkown\": \"arn:aws:kinesisvideo:us-west-2:123456789012:channel/test-channel/1234567890123\""        /* Unkown is ignored */
+    "}";
+    size_t messageLength = strlen( message );
+    SignalingChannelArn_t channelArn = { 0 };
+    SignalingResult_t result;
+
+    result = Signaling_ParseCreateSignalingChannelResponse( message,
+                                                            messageLength, &( channelArn ) );
+
+    TEST_ASSERT_EQUAL( SIGNALING_RESULT_OK,
```
Comment (ActoryOu) [inline test/unit-test/signaling_api/signaling_api_utest.c]: Does it make more sense to return `SIGNALING_RESULT_UNEXPECTED_RESPONSE` in this case?

Comment (Varunaditya1) [reply test/unit-test/signaling_api/signaling_api_utest.c]: As per my understanding if you get an known key then we are ignoring it in some responses, whereas we are giving `SIGNALING_RESULT_UNEXPECTED_RESPONSE` for other response.

Not sure how we are deciding which should do what !

File: source/signaling_api.c (hunk 5 above)
Comment (Varunaditya1) [reply source/signaling_api.c]: Yes, you are right. Will do the changes.

File: source/signaling_api.c (hunk 4 above)
Comment (Varunaditya1) [reply source/signaling_api.c]: Done in the latest commit.

**Review by ActoryOu — DISMISSED** [review]

**Review by aggarg — APPROVED** [review]

**Review by ActoryOu — APPROVED** [review]

**Pull request merged**